### PR TITLE
Add Expression Function to Read File

### PR DESCRIPTION
**Pull request by mfleader**
## Changes introduced with this PR

- [X] Add a function to the expression language to read a file into a string, and return that string
- [X] Add unit tests

---
By contributing to this repository, I agree to the [contribution guidelines](https://github.com/arcalot/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by jaredoconnell — APPROVED** [review]
Seems straightforward. It doesn't include any of our ideas for improving the security by pre-validating permission to access files outside of the working directory, but more prerequisites are required to make that work.

Let's see if Webb or Dave have any comments.

File: fixtures/test-readFile/hello-world.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+hello-world
```
Comment (webbnh) [inline fixtures/test-readFile/hello-world.yaml:1]: It is best if files end in a newline (unless you're deliberately testing the behavior when it's missing).

File: internal/builtinfunctions/functions.go [hunk 2]
```diff
@@ -490,3 +494,36 @@ func getSplitStringFunction() schema.CallableFunction {
 	}
 	return funcSchema
 }
+
+func getReadFileFunction() schema.CallableFunction {
+	funcSchema, err := schema.NewCallableFunction(
+		"readFile",
+		[]schema.Type{
+			schema.NewStringSchema(nil, nil, nil),
+		},
```
Comment (webbnh) [inline internal/builtinfunctions/functions.go:503]: This can all go on one line (see line 453).

**Review by webbnh — APPROVED** [review]
On it's face, the change looks good (except possibly for a nit or two).

I'm not sure what other constraints we want to place on file access.  Is there a canonical list of permitted/disallowed locations that we should be checking?

**Comment by mfleader** [comment]
> On its face, the change looks good (except possibly for a nit or two).
> 
> I'm not sure what other constraints we want to place on file access. Is there a canonical list of permitted/disallowed locations that we should be checking?

That's a great question.

**Comment by webbnh** [comment]
> > I'm not sure what other constraints we want to place on file access. Is there a canonical list of permitted/disallowed locations that we should be checking?
> 
> That's a great question. I do not know what we should disallow.

We cannot enforce a policy which we do not have.  Ship it!  (Unless you want to tweak the nits, first....)

File: fixtures/test-readFile/hello-world.yaml (hunk 1 above)
Comment (webbnh) [reply fixtures/test-readFile/hello-world.yaml]: Much better -- it looks much more like YAML now.  😉

**Review by webbnh — APPROVED** [review]
🚢

**Review by dbutenhof — APPROVED** [review]
This is cool. I can't help wondering if it'd be useful to be able to read an *object*, e.g. from a YAML file, perhaps even with reference to a schema for validation; but even if that's desirable, it can wait.

**Pull request merged**